File: swift/Sources/TypeSafeBuilder/Point3DBuilder.swift
public struct Point3DBuilder<X, Y, Z>
	where
		X: StaticOptional,
		X.Element == Double,
		Y: StaticOptional,
		Y.Element == Double,
		Z: StaticOptional,
		Z.Element == Double
{
	let x: X
	let y: Y
	let z: Z
}

public func point3d() -> Point3DBuilder<StaticNone<Double>,StaticNone<Double>,StaticNone<Double>> {
	Point3DBuilder(x: StaticNone(), y: StaticNone(), z: StaticNone())
}

extension Point3DBuilder
		where X == StaticNone<Double>
{
	public func x(_ x: Double) -> Point3DBuilder<StaticSome<Double>, Y, Z> {
		Point3DBuilder<StaticSome<Double>, Y, Z>(x: StaticSome(element: x), y: y, z: z)
	}
}

extension Point3DBuilder
	where Y == StaticNone<Double>
{
	public func y(_ y: Double) -> Point3DBuilder<X, StaticSome<Double>, Z> {
		Point3DBuilder<X, StaticSome<Double>, Z>(x: x, y: StaticSome(element: y), z: z)
	}
}

extension Point3DBuilder
	where Z == StaticNone<Double>
{
	public func z(_ z: Double) -> Point3DBuilder<X, Y, StaticSome<Double>> {
		Point3DBuilder<X, Y, StaticSome<Double>>(x: x, y: y, z: StaticSome(element: z))
	}
}

extension Point3DBuilder
	where
		X == StaticSome<Double>,
		Y == StaticSome<Double>,
		Z == StaticSome<Double>
{
	public func build() -> Point3D {
		Point3D(x: x.element, y: y.element, z: z.element)
	}
}

File: go/builder/point3DBuilder.go
package builder

import "type-safe-builder/point"

type Point3DBuilder[XType StaticOptional[float64], YType StaticOptional[float64], ZType StaticOptional[float64]] struct {
	x XType
	y YType
	z ZType
}

func Point3D() Point3DBuilder[StaticNone, StaticNone, StaticNone] {
	return Point3DBuilder[StaticNone, StaticNone, StaticNone]{
		x: StaticNone{},
		y: StaticNone{},
		z: StaticNone{},
	}
}

func X[YType StaticOptional[float64], ZType StaticOptional[float64]](builder Point3DBuilder[StaticNone, YType, ZType], x float64) Point3DBuilder[StaticSome[float64], YType, ZType] {
	return Point3DBuilder[StaticSome[float64], YType, ZType]{
		x: StaticSome[float64]{Value: x},
		y: builder.y,
		z: builder.z,
	}
}

func Y[XType StaticOptional[float64], ZType StaticOptional[float64]](builder Point3DBuilder[XType, StaticNone, ZType], y float64) Point3DBuilder[XType, StaticSome[float64], ZType] {
	return Point3DBuilder[XType, StaticSome[float64], ZType]{
		x: builder.x,
		y: StaticSome[float64]{Value: y},
		z: builder.z,
	}
}

func Z[XType StaticOptional[float64], YType StaticOptional[float64]](builder Point3DBuilder[XType, YType, StaticNone], z float64) Point3DBuilder[XType, YType, StaticSome[float64]] {
	return Point3DBuilder[XType, YType, StaticSome[float64]]{
		x: builder.x,
		y: builder.y,
		z: StaticSome[float64]{Value: z},
	}
}

func Build(builder Point3DBuilder[StaticSome[float64], StaticSome[float64], StaticSome[float64]]) point.Point3D {
	return point.Point3D{
		X: builder.x.Value,
		Y: builder.y.Value,
		Z: builder.z.Value,
	}
}
